### ISSUE TITLE
Would writing a lot of helpers functions be useful?

### ISSUE THREAD

**Issue by amoeba** [ropensci/EML#196]
This has probably been talked about in other Issues or off GH but I know at least @mbjones and I have talked about the need for this and we agree it's a good idea to provide a bunch of helper functions.

I see the main reason for helpers being to:

1. Speed the user up
2. Get away from `new` and `slots` which are confusing and clumsy

Plus, with `new` they don't get autocomplete and help docs. Maybe you're already well aware of all of this and you've just been hoping someone would do something about it.

I just looked over what's already there and there's a lot. We're already using `set_attributes` and `set_coverage` a lot but the first thing I found is that there are helpers like `set_temporalCoverage` that aren't exported. I'd like to make a few changes:

- Can we export more/all of these for the user?
- Can we tweak some of them to support more diverse use cases?

  For example, `set_coverage` doesn't support multiple geo coverages but that's a common use case. I'd have to sit down and think about what the interface would look like.
- Can we write lots more?

Here, 'we' can be 'me'. I just wanted to get thoughts before I code and send PR's.

**Comment by cboettig** [comment]
Couldn't agree more about helper functions, I think the real question is how to do it.  I suspect we should really be doing this in the standard S4 methods way rather than writing functions like `set_object` and `get_object`.  In most S4-based packages, users never call `new("individualName", givenName = "bob")`, but instead would create it with an initialize method whose name matches the object.  

```r
mbjones <- individualName(givenName = "Matt")
```

Likewise instead of a `get_` method, one would access slots with an accessor method based on the slot name:

```r
givenName(mbjones) # returns Matt
```

And similarly one would have assignment method, to assign a blank slot after initialization or to change a slot value:

```r
givenName(mbjones) <- "Matt"
surName(mbjones) <- "Jones"
```

The main reason I haven't done this is that it's simply a lot of writing -- there's 680 classes created in `classes.R`.  @gmbecker mentioned to me that he had been working on an S4 Coffee package that could generate these standard methods from the classes themselves, but I'm not sure it's out yet.  Automatically generating all these methods for all the classes would at least mean you should get he autocomplete and the help docs, which would be a big win.  Also following the S4 standard approach here should make the whole thing more intuitive at least to users familiar with S4 or other more object-oriented design (though I think most R users tend towards a function-driven design and focus on verbs like `set_` `get_`, rather than nouns, so I dunno.  Trying to remember some very good suggestions from @richfitz in this regard).  

So far I've been pretty conservative in what / how many `set_` and `get_` methods I export, as you've noticed.  My reasoning was:  I'm hoping for a high-level API that isn't overwhelming in it's complexity, but just makes it easy to do the "common" things, whatever those are.  I think this means not having too many functions, using sensible defaults, and just pointing advanced users to the lower-level API when they need special cases.  (Of course this means knowing what the special cases are -- e.g. maybe you're right that repeating geo elements are not a special case).  

The other reason I haven't exported more of the current `set_` `get_` methods is I'm not sure we're ready to commit to them in their current form and support them in the future -- some might be replaced by the S4 mechanism I described above, others could probably afford more fine-tuning to be more helpful in common cases.  

So, suggestion to go forward:

1.  it would probably be helpful for us to write a few S4 methods out by hand for common classes just to see how much we like it.  If we like that route, we then automate the creation of the methods for all the other classes (e.g. using S4 coffee or attempting our own go at this, like I've already done for some initializer methods to simplify the creation of repeating elements).  

2. New and tweaks to existing `set_` `get_` methods are welcome, but should emphasize ease of use over generality (with good design maybe we can do both).  These methods would also be easier to write once (1) is in place.

3. Ultimately we also want a better understanding of use cases and common patterns (maybe there's also an opportunity here to think about how to push users into creating the most useful EML).  

Thoughts?

**Comment by amoeba** [comment]
Excellent and helpful response. Thanks. My thoughts:

- I like and even prefer the S4 initializer/setter/getter approach outline 👍 
- There's too much in the EML spec to hand-write initializer/getter/setter methods so yes, a tool like S4 Coffee would be probably crucial to doing this
- Your goals with the `set_` and `_get` functions make sense

I'll close this (reopen if you prefer) and aim to write some S4 methods next time I'm working with helpers to get a feel for how I like it.

**Comment by cboettig** [comment]
Let's leave this issue open until we find a better home to document some of these design principles? (I don't think they are written out in one place anywhere else)  Could go in a CONTRIBUTING.md or somewhere else?  

We should also have issue(s)/milestone task for creating the S4 methods at least.

**Comment by amoeba** [comment]
Great, thanks. CONTRIBUTING.md would work well for these details. I might also add a note in the helpers themselves to avoid complaints about the generality.

**Comment by cboettig** [comment]
Great, closing in favor of #197.  Some general principles are now mentioned in CONTRIBUTING.md as well.
